### ISSUE TITLE
engine: Volumes stuck in termination stage

### ISSUE THREAD

**Issue by stiks** [longhorn/longhorn#259]
**Version:**
0.3-rc5

```
➜ kubectl -n longhorn-system delete volumes.longhorn.rancher.io --all

volume.longhorn.rancher.io "pvc-a12f7630-a6d6-11e8-a416-de2b4c2c400c" deleted
volume.longhorn.rancher.io "pvc-c93a8048-a6d7-11e8-a416-de2b4c2c400c" deleted
```
```
➜  kubectl -n longhorn-system delete volumes.longhorn.rancher.io --all

volume.longhorn.rancher.io "pvc-a12f7630-a6d6-11e8-a416-de2b4c2c400c" deleted
volume.longhorn.rancher.io "pvc-c93a8048-a6d7-11e8-a416-de2b4c2c400c" deleted
```
```
➜  kubectl -n longhorn-system describe volumes.longhorn.rancher.io

Name:         pvc-a12f7630-a6d6-11e8-a416-de2b4c2c400c
Namespace:    longhorn-system
Labels:       longhornvolume=pvc-a12f7630-a6d6-11e8-a416-de2b4c2c400c
Annotations:  <none>
API Version:  longhorn.rancher.io/v1alpha1
Kind:         Volume
Metadata:
  Creation Timestamp:             2018-08-23T13:15:47Z
  Deletion Grace Period Seconds:  0
  Deletion Timestamp:             2018-08-24T06:34:56Z
  Finalizers:
    longhorn.rancher.io
  Generation:        2
  Resource Version:  1048348
  Self Link:         /apis/longhorn.rancher.io/v1alpha1/namespaces/longhorn-system/volumes/pvc-a12f7630-a6d6-11e8-a416-de2b4c2c400c
  UID:               a5fd3fd9-a6d6-11e8-bb2a-de2b4c33000b
Spec:
  Base Image:
  Engine Image:        rancher/longhorn-engine:v0.3-rc5
  From Backup:
  Frontend:            blockdev
  Migration Node ID:
  Node ID:
  Number Of Replicas:  3
  Owner ID:            scw-4gb-ams1-node3
  Recurring Jobs:
    Cron:                 00 00 * * *
    Name:                 c-v53b7q
    Retain:               14
    Task:                 backup
  Size:                   5368709120
  Stale Replica Timeout:  30
Status:
  Conditions:
    Scheduled:
      Last Probe Time:
      Last Transition Time:  2018-08-23T13:15:48Z
      Message:
      Reason:
      Status:                True
      Type:                  scheduled
  Current Image:             rancher/longhorn-engine:v0.3-rc5
  Robustness:                unknown
  State:                     detached
Events:                      <none>


Name:         pvc-c93a8048-a6d7-11e8-a416-de2b4c2c400c
Namespace:    longhorn-system
Labels:       longhornvolume=pvc-c93a8048-a6d7-11e8-a416-de2b4c2c400c
Annotations:  <none>
API Version:  longhorn.rancher.io/v1alpha1
Kind:         Volume
Metadata:
  Creation Timestamp:             2018-08-23T13:24:02Z
  Deletion Grace Period Seconds:  0
  Deletion Timestamp:             2018-08-24T06:34:56Z
  Finalizers:
    longhorn.rancher.io
  Generation:        2
  Resource Version:  1048349
  Self Link:         /apis/longhorn.rancher.io/v1alpha1/namespaces/longhorn-system/volumes/pvc-c93a8048-a6d7-11e8-a416-de2b4c2c400c
  UID:               cd1bb1c6-a6d7-11e8-bb2a-de2b4c33000b
Spec:
  Base Image:
  Engine Image:        rancher/longhorn-engine:v0.3-rc5
  From Backup:
  Frontend:            blockdev
  Migration Node ID:
  Node ID:
  Number Of Replicas:  3
  Owner ID:            scw-4gb-ams1-node3
  Recurring Jobs:
    Cron:                 0 0 * * *
    Name:                 c-npgks9
    Retain:               5
    Task:                 backup
  Size:                   2147483648
  Stale Replica Timeout:  30
Status:
  Conditions:
    Scheduled:
      Last Probe Time:
      Last Transition Time:  2018-08-23T13:24:04Z
      Message:
      Reason:
      Status:                True
      Type:                  scheduled
  Current Image:             rancher/longhorn-engine:v0.3-rc5
  Robustness:                unknown
  State:                     detached
Events:                      <none>
```
```
➜  kubectl -n longhorn-system delete volumes.longhorn.rancher.io --all

volume.longhorn.rancher.io "pvc-a12f7630-a6d6-11e8-a416-de2b4c2c400c" deleted
volume.longhorn.rancher.io "pvc-c93a8048-a6d7-11e8-a416-de2b4c2c400c" deleted
```

I've tried even deleting longhorn, but now namespace stuck on deletion.

For thouse of you, who's having same issue.  This is how you can fix it:

```
➜  kubectl -n longhorn-system get volumes.longhorn.rancher.io pvc-a12f7630-a6d6-11e8-a416-de2b4c2c400c -o json > tt.json

{
    "apiVersion": "longhorn.rancher.io/v1alpha1",
    "kind": "Volume",
    "metadata": {
        "creationTimestamp": "2018-08-23T13:15:47Z",
        "deletionGracePeriodSeconds": 0,
        "deletionTimestamp": "2018-08-24T06:37:02Z",
        "finalizers": [
            "longhorn.rancher.io" .      <-- REMOVE THIS LINE
        ],
        "generation": 2,
        "labels": {
            "longhornvolume": "pvc-a12f7630-a6d6-11e8-a416-de2b4c2c400c"
        },
        "name": "pvc-a12f7630-a6d6-11e8-a416-de2b4c2c400c",
        "namespace": "longhorn-system",
        "resourceVersion": "1048713",
        "selfLink": "/apis/longhorn.rancher.io/v1alpha1/namespaces/longhorn-system/volumes/pvc-a12f7630-a6d6-11e8-a416-de2b4c2c400c",
        "uid": "a5fd3fd9-a6d6-11e8-bb2a-de2b4c33000b"
    },
    "spec": {
        "baseImage": "",
        "engineImage": "rancher/longhorn-engine:v0.3-rc5",
        "fromBackup": "",
        "frontend": "blockdev",
        "migrationNodeID": "",
        "nodeID": "",
        "numberOfReplicas": 3,
        "ownerID": "scw-4gb-ams1-node3",
        "recurringJobs": [
            {
                "cron": "00 00 * * *",
                "name": "c-v53b7q",
                "retain": 14,
                "task": "backup"
            }
        ],
        "size": "5368709120",
        "staleReplicaTimeout": 30
    },
    "status": {
        "conditions": {
            "scheduled": {
                "lastProbeTime": "",
                "lastTransitionTime": "2018-08-23T13:15:48Z",
                "message": "",
                "reason": "",
                "status": "True",
                "type": "scheduled"
            }
        },
        "currentImage": "rancher/longhorn-engine:v0.3-rc5",
        "robustness": "unknown",
        "state": "detached"
    }
}
```
Save it to new file, then push it back into that volume:
```
curl -H "Content-Type: application/json" -X PUT --data-binary @tt.json http://127.0.0.1:8001/apis/longhorn.rancher.io/v1alpha1/namespaces/longhorn-system/volumes/pvc-a12f7630-a6d6-11e8-a416-de2b4c2c400c
```
I have to do it couple time, when volume finally gets deleted.
```
➜  kubectl -n longhorn-system get volumes.longhorn.rancher.io
No resources found.

➜  kubectl -n longhorn-system get engines.longhorn.rancher.io
NAME                                                  AGE
pvc-a12f7630-a6d6-11e8-a416-de2b4c2c400c-e-f5b48ed0   17h
pvc-c93a8048-a6d7-11e8-a416-de2b4c2c400c-e-9ee1829c   17h

➜  kubectl -n longhorn-system get replicas.longhorn.rancher.io
NAME                                                  AGE
pvc-a12f7630-a6d6-11e8-a416-de2b4c2c400c-r-333b1c13   15h
pvc-a12f7630-a6d6-11e8-a416-de2b4c2c400c-r-b880e025   17h
pvc-a12f7630-a6d6-11e8-a416-de2b4c2c400c-r-c0db5c25   15h
pvc-c93a8048-a6d7-11e8-a416-de2b4c2c400c-r-4a31b2bb   15h
pvc-c93a8048-a6d7-11e8-a416-de2b4c2c400c-r-dc68ec61   17h
pvc-c93a8048-a6d7-11e8-a416-de2b4c2c400c-r-fb03d09a   15h
```

Longhorn manager in mean time doing this:
```
E0824 06:49:23.913824 1 reflector.go:205] github.com/rancher/longhorn-manager/vendor/k8s.io/client-go/informers/factory.go:87: Failed to list *v1.Job: Unauthorized
E0824 06:49:23.938584 1 reflector.go:205] github.com/rancher/longhorn-manager/vendor/k8s.io/client-go/informers/factory.go:87: Failed to list *v1beta1.CronJob: Unauthorized
E0824 06:49:23.980517 1 reflector.go:205] github.com/rancher/longhorn-manager/vendor/k8s.io/client-go/informers/factory.go:87: Failed to list *v1.Pod: Unauthorized
E0824 06:49:23.983570 1 reflector.go:205] github.com/rancher/longhorn-manager/k8s/pkg/client/informers/externalversions/factory.go:74: Failed to list *v1alpha1.EngineImage: Unauthorized
E0824 06:49:24.008624 1 reflector.go:205] github.com/rancher/longhorn-manager/vendor/k8s.io/client-go/informers/factory.go:87: Failed to list *v1beta2.DaemonSet: Unauthorized
```

**Comment by stiks** [comment]
Even after all this magic, all 3 volumes still present and stuck:

```
{
"accessModes": [
"ReadWriteOnce"
],
"annotations": {
"pv.kubernetes.io/provisioned-by": "rancher.io/longhorn"
},
"awsElasticBlockStore": null,
"azureDisk": null,
"azureFile": null,
"capacity": {
"storage": "5Gi"
},
"cephfs": null,
"cinder": null,
"claimRef": {
"apiVersion": "v1",
"kind": "PersistentVolumeClaim",
"name": "grafana-grafana",
"namespace": "monitoring",
"resourceVersion": "860054",
"type": "/v3/cluster/schemas/objectReference",
"uid": "a12f7630-a6d6-11e8-a416-de2b4c2c400c"
},
"created": "2018-08-23T13:15:47Z",
"creatorId": null,
"csi": {
"driver": "io.rancher.longhorn",
"readOnly": false,
"type": "/v3/cluster/schemas/csiPersistentVolumeSource",
"volumeAttributes": {
"fromBackup": "",
"numberOfReplicas": "3",
"staleReplicaTimeout": "30",
"storage.kubernetes.io/csiProvisionerIdentity": "1535025566987-8081-rancher.io/longhorn"
},
"volumeHandle": "pvc-a12f7630-a6d6-11e8-a416-de2b4c2c400c"
},
"fc": null,
"flexVolume": null,
"flocker": null,
"gcePersistentDisk": null,
"glusterfs": null,
"hostPath": null,
"iscsi": null,
"local": null,
"mountOptions": [ ],
"name": "pvc-a12f7630-a6d6-11e8-a416-de2b4c2c400c",
"nfs": null,
"nodeAffinity": null,
"ownerReferences": [ ],
"persistentVolumeReclaimPolicy": "Delete",
"photonPersistentDisk": null,
"portworxVolume": null,
"quobyte": null,
"rbd": null,
"removed": "2018-08-23T23:08:46Z",
"scaleIO": null,
"state": "removing",
"status": {
"phase": "Released",
"type": "/v3/cluster/schemas/persistentVolumeStatus"
},
"storageos": null,
"transitioning": "error",
"transitioningMessage": "waiting on external-attacher/io-rancher-longhorn",
"uuid": "a61544de-a6d6-11e8-a416-de2b4c2c400c",
"vsphereVolume": null
}
```

**Comment by stiks** [comment]
I've deleted previous longhorn install (which was stuck on deleting) and installed it once again. It did start, 2 nodes of 5 keep crashing. In a couple hours all nodes appear to have issues with permission. It may be not related to the stuck issue (I think kubernetes deleted roles)

**Comment by yasker** [comment]
@stiks If there is not enough permission for Longhorn to operate on Kubernetes resources, it won't be able to clean up the resource when volumes are deleted. We've put finalizer in the Kubernetes objects to make sure it was cleaned up properly before removed. That's why the deleted volume won't disappear, because managers cannot clean up after it without Kubernetes permission.

**Comment by yasker** [comment]
@stiks Also, since there are finalizers in place for resource cleanup, you would need to follow the uninstall instruction to delete the Longhorn, otherwise, it won't be able to complete deletion correct (get stuck).

Though we should mention that you need to delete PVC/PV in the Kubernetes first, before cleanup the Longhorn. I will update the uninstall instruction.

**Comment by stiks** [comment]
I did delete all PVC before. This is what I did.

I've replaced 2 nodes in the cluster. Added few extra disks. Longhorn kept previous nodes in the list even they no longer exist in the cluster. 3 PVC I have had on these nodes has been moved to new instances, but I have had issues with them. I've decided to redo them. This is where volumes get stuck, I cannot longer delete them. Even when I did all that magic with finalizers, PVC has been stuck and don't want to be deleted. I tried deleting longhorn, but that didn't helped

**Comment by yasker** [comment]
@stiks Sometime Kubernetes is stubborn...

Longhorn is definitely not in a working state in your system right now. Can you try to reinstall it using the latest instruction? I think PVC is trying to communicate with Longhorn but cannot find it, so it's stuck there. And previously because Longhorn somehow lost the permission to work with Kubenetes, so it cannot work. That's maybe why PVC stuck in the first place.

**Comment by stiks** [comment]
Yeah, in the end, so many things were stuck, couldn't move anywhere. I've to rebuild cluster. Now I'm on 0.3, will let you know if there'll be any issues with setup.

**Comment by ppotaki** [comment]
I reproduced this issue on my Rancher Cluster - the volumes were in Terminating state.

> pvc-1822b0b0e53f11e8                       10Gi       RWO            Delete           Terminating   volume-test/pvc-6gflk           sc-4n6v6                    3d
> pvc-1f7ac678e54711e8                       5Gi        RWO            Delete           Terminating   default/pvc-vbdpj                     sc-4n6v6                    3d
> pvc-21581baa-e08b-11e8-aca0-000c2937d4e3   10Gi       RWO            Delete           Terminating   volume-test/volume-test   longhorn                    3d
> 

I was able to delete volumes from Rancher interface by editing PVC  with kubectl and deleting following **finalizer** section from yaml for above PVC's. I hope it will help you with your cases.

**Comment by yasker** [comment]
@ppotaki Yeah, remove finalizer would work but it may result in leftover data on the node. How did you get into this state? If you deleted the Longhorn before the volumes, it will behave like this.

Now you can follow the uninstallation instructions [here](https://github.com/rancher/longhorn#uninstall-longhorn) to uninstall Longhorn properly.

**Comment by josecelano** [comment]
Same problem. I have created a new issue so people can find the issue faster:

https://github.com/longhorn/longhorn/issues/722

I have not found a solution yet. I followed the [uninstallation instructions](https://github.com/longhorn/longhorn#uninstall-longhorn) here to uninstall Longhorn properly.

Related issue:
https://github.com/longhorn/longhorn/issues/344

**Comment by josecelano** [comment]
I had to remove the finalizers using rancher `kubectl` console because even if the PVs were deleted from GKE they still were present in rancher, so I guess rancher has an internal cache and was not able to sync state with GKE.

https://github.com/longhorn/longhorn/issues/722

**Comment by yasker** [comment]
@josecelano I think GKE is the problem (and has the cache). Rancher reads all the information from `kubectl` which is a standard Kubernetes interface to the Kubernetes API. Rancher didn't control the result of `kubectl`. If `kubectl` say the PV still exists but GKE think it's gone, then GKE is wrong.
